### ISSUE TITLE
Unable to get url of s3 file

### ISSUE THREAD

**Issue by martinsefcik** [janko-m/shrine#54]
The following call _record.file.url_ fails in case metadata[id] contains space

```
bad URI(is not URI?): invoice/2016/sample file.pdf
...
"/var/lib/gems/2.2.0/gems/shrine-1.3.0/lib/shrine/storage/s3.rb:194:in `url'"
...
```

It looks like file is uploaded correctly but I'm not able to get its url.
Maybe id should be **Uri.encode**d before joining with host.

**Comment by janko-m** [comment]
Thanks for reporting. Yeah, it would probably be much better to generate a `public_url`, and then just take the path part to join with the `:host`. I will push a fix shortly.

**Comment by martinsefcik** [comment]
Thanks!

**Comment by martinsefcik** [comment]
I'm not sure if this was correctly fixed, because now url contains host and bucket name together (https://host.com/bucket_name/path/file.ext) and this is not correct. It should be https://host.com/path/file.ext in case host is defined.

**Comment by janko-m** [comment]
Hmm, are you by any chance instantiating the storage with `force_path_style: true`? Because this bug would happen if your S3 URLs look like http://s3.amazonaws.com/bucket-name/foo.jpg instead of http://bucket-name.s3.amazonaws.com/foo.jpg. I'm just wondering if the former can ever happen without `force_path_style: true`.

**Comment by janko-m** [comment]
I went with your original advice of just wrapping `URI.encode` around the object key. I think this should work good now.

**Comment by martinsefcik** [comment]
I'm not using ```force_path_style``` option at all. I'm using cloudflare CDN with own domain so ```bucket_name``` is not in url.
I tested it with d1a350a3093a5ea8b60226980c2bcc8a27aba03b and it's working fine now.

**Comment by janko-m** [comment]
@martinsefcik I'm trying to figure out why does your presigned URLs return bucket name in the URI path instead of the host. For me it always returns bucket name in the URI host. I don't like this nondeterminism, and I would like to report it to the aws-sdk team, but I don't know how to reproduce it. Maybe it is different for different AWS regions?

**Comment by martinsefcik** [comment]
Currently I have no issue with 1.4.2 version. As I mentioned it was fixed in d1a350a3093a5ea8b60226980c2bcc8a27aba03b.

Before that commit there was a problem. In case custom host (CDN) was defined then bucket_name was added to the URI path.

**Comment by janko-m** [comment]
When you run
```rb
Shrine.storages[:store].bucket.object("foo").presigned_url(:get)
Shrine.storages[:store].bucket.object("foo").public_url
```
do they return S3 URLs with bucket name in the _host_ or in the _path_? If you got that error, this should return for you the URL with bucket name in the _path_. In that case there is something different in our configuration, because I always get the bucket name in the URI host (I tried with every region).
